### PR TITLE
Add stack backtraces and screenshot/json for REST API failures

### PR DESCRIPTION
**Pull request by jknphy**
See [poo#89482](https://progress.opensuse.org/issues/89482)
Verification run: [SLE](https://openqa.suse.de/tests/overview.html?distri=sle&build=jknphy%2Fos-autoinst-distri-opensuse%23improve_post_fail_rest_api&version=15-SP3) | [openSUSE-Tumbleweed-x86_64](https://openqa.opensuse.org/tests/latest?arch=x86_64&distri=opensuse&flavor=DVD&machine=64bit&test=ext4%40jknphy%2Fos-autoinst-distri-opensuse%23improve_post_fail_rest_api&version=Tumbleweed#live) | [opensuse-Tumbleweed-DVD-ppc64le](https://openqa.opensuse.org/tests/latest?arch=ppc64le&distri=opensuse&flavor=DVD&machine=ppc64le&test=RAID0%40jknphy%2Fos-autoinst-distri-opensuse%23improve_post_fail_rest_api&version=Tumbleweed) | [opensuse-15.3](https://openqa.opensuse.org/tests/latest?arch=x86_64&distri=opensuse&flavor=DVD&machine=64bit&test=RAID0_gpt%40jknphy%2Fos-autoinst-distri-opensuse%23improve_post_fail_rest_api&version=15.3#live)

### CODE REVIEW

File: lib/y2_rest_api.pm [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+=cut
+
+sub post_fail_hook {
+    my $self = shift;
+    save_screenshot;
+    if (get_var('YUI_REST_API')) {
+        my $json = to_json(YuiRestClient::connect_to_running_app());
+        save_tmp_file('widgets.json', $json);
+        select_console('root-console');
```
Comment (rwx788) [inline lib/y2_rest_api.pm:35]: I wonder if we should use `log-console`, also keep in mind that we can do this request from the worker directly and print things to the logs. We can use same mechanism, as we do for autoyast profiles, see`sub upload_profile`.

**Review by rwx788 — COMMENTED** [review]
Couple of minor things, good job overall!

File: lib/y2_rest_api.pm [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved. This file is offered as-is,
+# without any warranty.
+
+# Maintainer: QE YaST <qa-sle-yast@suse.de>
+
+package y2_rest_api;
+
+use base "y2_installbase";
```
[the remark this answers is not in the document]
Comment (jknphy) [reply lib/y2_rest_api.pm:14]: I was not sure how to covert that, I can insert y2_rest_api between y2_installbase and installbasetest in the hierarchy, but for the modules we have y2_module_basetest inheriting from opensusebasetest, but then I would be using installbasetest for modules which doesn't make sense. Perhaps I didn't understood your proposal.

File: lib/y2_rest_api.pm (hunk 1 above)
Comment (jknphy) [reply lib/y2_rest_api.pm]: good idea, I just understood how this ulogs magic folder works, makes sense.

File: lib/y2_rest_api.pm (hunk 2 above)
[the remark this answers is not in the document]
Comment (rwx788) [reply lib/y2_rest_api.pm:14]: Ah, true. We could use mutli inheritance which is supported in perl. I need to check what we have to make a proposal, but we definitely need same thing for tests in the installed system.

[the remark this answers is not in the document]
Comment (rwx788) [reply lib/y2_rest_api.pm:14]: After a quick look, as `installbasetest` just sets fatal flag, we can move it to `y2_installbase` and then we can inherit from `y2_rest_api`. With yast modules it's easier, `y2_module_basetest` inherits from `opensusebasetest`. So we can inherit from ``y2_rest_api`` and ``y2_rest_api`` will have `opensusebasetest` as a base class.
As an alternative we have multiple inheritance. Also, I believe that we can just use import where needed, as in `y2_module_basetest` we already call methods from `y2_installbase`, which means we had a need for a better structure before.

[the remark this answers is not in the document]
Comment (jknphy) [reply lib/y2_rest_api.pm:14]: I got  early today to the same setup you proposed, thanks for taking a look and confirming. Trying to run it to see all is fine.

**Comment by okurz** [comment]
s/screenshoot/screenshot/

File: lib/y2_rest_api.pm (hunk 2 above)
[the remark this answers is not in the document]
Comment (jknphy) [reply lib/y2_rest_api.pm]: done!

File: tests/x11/network/yast2_network_use_nm.pm [hunk 3]
```diff
@@ -14,7 +14,7 @@
 
 use base 'x11test';
 use y2_module_guitest 'launch_yast2_module_x11';
-use y2_installbase;
+use y2_base;
```
Comment (rwx788) [inline tests/x11/network/yast2_network_use_nm.pm:17]: This is very strange, what we do here, `y2_module_guitest` should include post fail hook, as for ncurses we have `y2_module_consoletest`. As it's out of scope here, could you please file a ticket to make sure we have correct structure being used?

**Review by rwx788 — APPROVED** [review]
Neat!

File: tests/x11/network/yast2_network_use_nm.pm (hunk 3 above)
Comment (jknphy) [reply tests/x11/network/yast2_network_use_nm.pm:17]: https://progress.opensuse.org/issues/91761

**Pull request merged**